File: ios/Vibree/Vibree/Utils/StressAnalysisAgent.swift
import Foundation

/**
 * Edge AI Agent for local stress analysis on iOS.
 * Integrates PyTorch Mobile (CNN-LSTM and MobileBERT) with heuristic fallback.
 */
class StressAnalysisAgent {
    static let shared = StressAnalysisAgent()
    
    private var rrBuffer: [Int] = []
    private let BUFFER_SIZE = 60
    
    private var hrvModule: TorchModule?
    private var nlpModule: TorchModule?
    private let tokenizer = BertTokenizer()
    
    enum StressLabel: String {
        case relaxed = "Relaxed"
        case lowStress = "Low Stress"
        case mediumStress = "Medium Stress"
        case highStress = "High Stress"
        case unknown = "Unknown"
    }
    
    struct Prediction {
        let stressScore: Int
        let label: StressLabel
        let confidence: Float
        let rmssd: Double
        let sdnn: Double
    }
    
    private init() {
        if let hrvPath = Bundle.main.path(forResource: "hrv_model", ofType: "ptl") {
            hrvModule = TorchModule(fileAtPath: hrvPath)
        }
        if let nlpPath = Bundle.main.path(forResource: "nlp_model", ofType: "ptl") {
            nlpModule = TorchModule(fileAtPath: nlpPath)
        }
    }
    
    func process(rrIntervalMs: Int) -> Prediction {
        if rrIntervalMs <= 0 || rrIntervalMs > 2000 {
            return emptyPrediction()
        }
        
        updateBuffer(rr: rrIntervalMs)
        
        if rrBuffer.count < 10 {
            return emptyPrediction()
        }
        
        let rmssd = calculateRMSSD(data: rrBuffer)
        let sdnn = calculateSDNN(data: rrBuffer)
        
        var mlScore: Int = -1
        if let module = hrvModule {
            // Construct features: [34 features] matches CNN-LSTM training data
            var features = [Double](repeating: 0.0, count: 34)
            features[3] = rmssd
            features[4] = sdnn
            
            if let output = module.predictPayload(features as [NSNumber]) {
                if output.count >= 2 {
                    let nonStress = output[0].floatValue
                    let stress = output[1].floatValue
                    
                    if stress > nonStress {
                         let diff = stress - nonStress
                         mlScore = 50 + Int(min(diff * 5, 50))
                    } else {
                         let diff = nonStress - stress
                         mlScore = 50 - Int(min(diff * 5, 50))
                    }
                }
            }
        }
        
        return inferStress(rmssd: rmssd, sdnn: sdnn, mlScore: mlScore)
    }
    
    /**
     * Analyze context for stress markers.
     */
    func analyzeText(_ text: String) -> Int {
        let (ids, mask) = tokenizer.tokenize(text)
        
        if let module = nlpModule, let output = module.predictNLPWithIds(ids as [NSNumber], mask: mask as [NSNumber]) {
            if output.count >= 2 {
                let nonStress = output[0].floatValue
                let stress = output[1].floatValue
                if stress > nonStress {
                    return 70 + Int(min((stress - nonStress) * 10, 30))
                } else {
                    return 30 - Int(min((nonStress - stress) * 10, 30))
                }
            }
        }
        return -1
    }
    
    private func updateBuffer(rr: Int) {
        rrBuffer.append(rr)
        if rrBuffer.count > BUFFER_SIZE {
            rrBuffer.removeFirst()
        }
    }
    
    private func calculateRMSSD(data: [Int]) -> Double {
        var sumSquares = 0.0
        for i in 0..<(data.count - 1) {
            let diff = Double(data[i + 1] - data[i])
            sumSquares += pow(diff, 2.0)
        }
        return sqrt(sumSquares / Double(data.count - 1))
    }
    
    private func calculateSDNN(data: [Int]) -> Double {
        let total = data.reduce(0, +)
        let mean = Double(total) / Double(data.count)
        
        var sumSquaredDiff = 0.0
        for x in data {
            sumSquaredDiff += pow(Double(x) - mean, 2.0)
        }
        return sqrt(sumSquaredDiff / Double(data.count))
    }
    
    private func inferStress(rmssd: Double, sdnn: Double, mlScore: Int = -1) -> Prediction {
        var score = 0
        var label = StressLabel.unknown
        var confidence: Float = 0.8
        
        if rmssd < 20 {
            score = 80 + Int(min((20.0 - rmssd) / 20.0 * 20.0, 20.0))
        } else if rmssd < 40 {
            score = 50 + Int(min((40.0 - rmssd) / 20.0 * 30.0, 30.0))
        } else if rmssd < 70 {
            score = 20 + Int(min((70.0 - rmssd) / 30.0 * 30.0, 30.0))
        } else {
            score = max(0, 20 - Int(min((rmssd - 70.0) / 100.0 * 20.0, 20.0)))
        }
        
        if mlScore != -1 {
            score = Int(Double(score) * 0.4 + Double(mlScore) * 0.6)
            confidence = 0.95
        }
        
        if score >= 80 { label = .highStress }
        else if score >= 50 { label = .mediumStress }
        else if score >= 20 { label = .lowStress }
        else { label = .relaxed }
        
        let sizeFactor = Float(rrBuffer.count) / Float(BUFFER_SIZE)
        confidence *= min(sizeFactor, 1.0)
        
        return Prediction(stressScore: score, label: label, confidence: confidence, rmssd: rmssd, sdnn: sdnn)
    }
    
    private func emptyPrediction() -> Prediction {
        return Prediction(stressScore: 0, label: .unknown, confidence: 0.0, rmssd: 0.0, sdnn: 0.0)
    }
    
    func reset() {
        rrBuffer.removeAll()
    }
}

File: android/app/src/main/java/com/example/wearablecollector/logic/StressAnalysisAgent.kt
package com.example.wearablecollector.logic

import android.content.Context
import org.pytorch.IValue
import org.pytorch.LiteModuleLoader
import org.pytorch.Module
import org.pytorch.Tensor
import java.io.File
import java.io.FileOutputStream
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * Edge AI Agent for local stress analysis.
 * Uses PyTorch Mobile for deep learning inference combined with heuristics.
 */
object StressAnalysisAgent {
    private val rrBuffer = ArrayDeque<Int>()
    private const val BUFFER_SIZE = 60 // 60 beats sliding window (~1 minute)
    
    private var hrvModel: Module? = null
    private var nlpModel: Module? = null
    private var tokenizer: BertTokenizer? = null
    
    data class Prediction(
        val stressScore: Int, // 0-100
        val label: StressLabel,
        val confidence: Float, // 0.0 - 1.0
        val rmssd: Double,
        val sdnn: Double
    )

    enum class StressLabel {
        RELAXED, LOW_STRESS, MEDIUM_STRESS, HIGH_STRESS, UNKNOWN
    }
    
    fun initialize(context: Context) {
        try {
            val hrvPath = assetFilePath(context, "hrv_model.ptl")
            val nlpPath = assetFilePath(context, "nlp_model.ptl")
            if (hrvPath != null) {
                hrvModel = LiteModuleLoader.load(hrvPath)
            }
            if (nlpPath != null) {
                nlpModel = LiteModuleLoader.load(nlpPath)
            }
            tokenizer = BertTokenizer(context)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun assetFilePath(context: Context, assetName: String): String? {
        val file = File(context.filesDir, assetName)
        if (file.exists() && file.length() > 0) {
            return file.absolutePath
        }
        try {
            context.assets.open(assetName).use { inputStream ->
                FileOutputStream(file).use { os ->
                    val buffer = ByteArray(4 * 1024)
                    var read: Int
                    while (inputStream.read(buffer).also { read = it } != -1) {
                        os.write(buffer, 0, read)
                    }
                    os.flush()
                }
                return file.absolutePath
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    fun process(rrIntervalMs: Int): Prediction {
        if (rrIntervalMs <= 0 || rrIntervalMs > 2000) return emptyPrediction()

        updateBuffer(rrIntervalMs)

        if (rrBuffer.size < 10) return emptyPrediction() // Need minimum data

        val rmssd = calculateRMSSD(rrBuffer)
        val sdnn = calculateSDNN(rrBuffer)
        
        var mlScore = -1
        hrvModel?.let { model ->
            try {
                // Construct input tensor of shape (1, 1, 34) matching training HRV model
                val features = FloatArray(34) { 0f }
                features[3] = rmssd.toFloat()
                features[4] = sdnn.toFloat()
                val inputTensor = Tensor.fromBlob(features, longArrayOf(1, 1, 34))
                
                val outputTensor = model.forward(IValue.from(inputTensor)).toTensor()
                val scores = outputTensor.dataAsFloatArray
                
                if (scores.size >= 2) {
                    if (scores[1] > scores[0]) {
                        val diff = scores[1] - scores[0]
                        mlScore = 50 + (diff * 5).toInt().coerceIn(0, 50) 
                    } else {
                        val diff = scores[0] - scores[1]
                        mlScore = 50 - (diff * 5).toInt().coerceIn(0, 50)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        
        return inferStress(rmssd, sdnn, mlScore)
    }

    /**
     * Analyzes text for psychological stress markers using MobileBERT.
     */
    fun analyzeText(text: String): Int {
        val t = tokenizer ?: return -1
        val (ids, mask) = t.tokenize(text)
        return processNLP(ids, mask)
    }
    
    private fun processNLP(textTokens: LongArray, attentionMask: LongArray): Int {
        var nlpScore = -1
        nlpModel?.let { model ->
            try {
                val inputIdsTensor = Tensor.fromBlob(textTokens, longArrayOf(1, 128))
                val maskTensor = Tensor.fromBlob(attentionMask, longArrayOf(1, 128))
                
                val outputTensor = model.forward(IValue.from(inputIdsTensor), IValue.from(maskTensor)).toTensor()
                val scores = outputTensor.dataAsFloatArray
                
                if (scores.size >= 2) {
                    // Normalize to 0-100 scale [Non-Stress, Stress]
                    if (scores[1] > scores[0]) {
                        nlpScore = 70 + (scores[1] - scores[0]).toInt().coerceIn(0, 30)
                    } else {
                        nlpScore = 30 - (scores[0] - scores[1]).toInt().coerceIn(0, 30)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return nlpScore
    }

    private fun updateBuffer(rr: Int) {
        rrBuffer.addLast(rr)
        if (rrBuffer.size > BUFFER_SIZE) {
            rrBuffer.removeFirst()
        }
    }

    private fun calculateRMSSD(data: List<Int>): Double {
        var sumSquares = 0.0
        for (i in 0 until data.size - 1) {
            val diff = (data[i + 1] - data[i]).toDouble()
            sumSquares += diff.pow(2)
        }
        return sqrt(sumSquares / (data.size - 1))
    }

    private fun calculateSDNN(data: List<Int>): Double {
        val mean = data.average()
        var sumSquaredDiff = 0.0
        for (x in data) {
            sumSquaredDiff += (x - mean).pow(2)
        }
        return sqrt(sumSquaredDiff / data.size)
    }

    private fun inferStress(rmssd: Double, sdnn: Double, mlScore: Int = -1): Prediction {
        var score = 0
        var label = StressLabel.UNKNOWN
        var confidence = 0.8f 

        if (rmssd < 20) {
            score = 80 + ((20 - rmssd) / 20 * 20).toInt()
        } else if (rmssd < 40) {
            score = 50 + ((40 - rmssd) / 20 * 30).toInt()
        } else if (rmssd < 70) {
            score = 20 + ((70 - rmssd) / 30 * 30).toInt()
        } else {
            score = (20 - (rmssd - 70) / 100 * 20).toInt().coerceAtLeast(0)
        }
        
        if (mlScore != -1) {
             score = (score * 0.4 + mlScore * 0.6).toInt().coerceIn(0, 100)
             confidence = 0.95f
        }
        
        if (score >= 80) label = StressLabel.HIGH_STRESS
        else if (score >= 50) label = StressLabel.MEDIUM_STRESS
        else if (score >= 20) label = StressLabel.LOW_STRESS
        else label = StressLabel.RELAXED

        confidence *= (rrBuffer.size.toFloat() / BUFFER_SIZE).coerceAtMost(1.0f)

        return Prediction(score, label, confidence, rmssd, sdnn)
    }

    private fun emptyPrediction() = Prediction(0, StressLabel.UNKNOWN, 0f, 0.0, 0.0)

    fun reset() {
        rrBuffer.clear()
    }
}
